packages/@react-stately/color/src/useHexColorFieldState.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {Color} from './Color';
+import {HexColorFieldProps} from '@react-types/color';
+import {NumberFieldState} from '@react-stately/numberfield';
+import {useCallback, useState} from 'react';
+import {useColor} from './useColor';
+import {useControlledState} from '@react-stately/utils';
+
+export interface HexColorFieldState extends Omit<NumberFieldState, 'value' | 'setValue'> {
+  colorValue: Color,
+  setInputValue: (value: string) => void
+}
+
+export const MIN_COLOR = new Color('#000000');
+export const MAX_COLOR = new Color('#FFFFFF');
+const MIN_COLOR_INT = MIN_COLOR.toHexInt();
+const MAX_COLOR_INT = MAX_COLOR.toHexInt();
+
+export function useHexColorFieldState(
+  props: HexColorFieldProps
+): HexColorFieldState {
+  let {
+    step = 1,
+    value,
+    defaultValue,
+    onChange,
+    validationState
+  } = props;
+  
+  let {color: initialValue} = useColor(value);
+  let {color: initialDefaultValue} = useColor(defaultValue);
+  let [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
+
+  let initialInputValue = (value || defaultValue) && colorValue ? colorValue.toString('hex') : '';
+  let [inputValue, setInputValue] = useState(initialInputValue);
```

Comment:
Hi @devongovett, I'm not sure I understand your comment 😅.

In cases where changes in the input value lead the `value` change (like typing), the input doesn't update (`commitInputValue`) until you blur out of the field, so the input doesn't disrupt you and change the input text to uppercase when you're typing in lowercase.

In cases where you don't type (increment, decrement via keyboard/mouse), the input value already updates along with the change in color value

Reply:
I mean when the `value` prop updates from the outside:

```jsx
// Before
<HexColorField value="#abcdef" />

// After
<HexColorField value="#123456" />
```

I believe right now the input value would still stay as `#abcdef` after the prop change because the `inputValue` state is not updated.

Reply:
Here's how I handled it in numberfield, i ran into this one as well https://github.com/adobe/react-spectrum/pull/1029/files#diff-f920ff47503335092fb8b412b9e4d0679256fecedc505827ad70a7c7871354dbR104
and here's the test i wrote for it
https://github.com/adobe/react-spectrum/pull/1029/files#diff-89757429f0981d25f225267128d9551d5dee758c05de1c49d2ae1dd09fcb7306R1063

Reply:
Hi @devongovett @snowystinger,

I'm now able to have the input updated when a new `value` prop is passed by adding the snippet below in stately
```ts
  useEffect(() => {
    setInputValue(colorValue ? colorValue.toString('hex') : '');
  }, [colorValue, setInputValue]);
```

But now it's updating the input unnecessarily (I think). It reformats the string as you delete characters coz they are still valid color strings.
![d](https://user-images.githubusercontent.com/5743833/97055778-5a2c3600-153c-11eb-8698-3083d32066a9.gif)

Reply:
Ah good catch! Maybe you could only update the inputValue if the color coming from props is different from the one that's already in state. Something like this (untested):

```jsx
useEffect(() => {
  setInputValue(inputValue => {
    // Parse color from current inputValue.
    // Only update the input value if the new colorValue is not equivalent.
    let currentColor = new Color(inputValue);
    if (currentColor?.toHexInt() !== colorValue?.toHexInt()) {
      return colorValue ? colorValue.toString('hex') : '';
    }

    return inputValue;
  });
}, [colorValue, setInputValue]);
```

Reply:
Hi @devongovett, I've committed changes to resolve this issue, but I'm having issues writing the test for it.

I wrote the test below
```js
  it('should update value in controlled state', function () {
    function HexColorFieldControlled(props) {
      let {onChange} = props;
      let [color, setColor] = useState(props.value);
      return (
        <HexColorField
          {...props}
          label="Primary Color"
          value={color}
          onChange={chain(setColor, onChange)} />
      ); 
    }
    let onChangeSpy = jest.fn();
    let {getByLabelText} = render(<HexColorFieldControlled value={new Color('#abc')} onChange={onChangeSpy} />);

    let hexColorField = getByLabelText('Primary Color');
    expect(hexColorField.value).toBe('#AABBCC');

    act(() => {
      hexColorField.focus();
      userEvent.clear(hexColorField);
    });
    // should call onChange when input is cleared
    expect(hexColorField.value).toBe('');
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
    expect(onChangeSpy).toHaveBeenCalledWith(null);

    typeText(hexColorField, 'cba');
    expect(hexColorField.value).toBe('cba');
    expect(onChangeSpy).toHaveBeenCalledTimes(2);
    expect(onChangeSpy).toHaveBeenCalledWith(new Color('#cba'));

    act(() => {hexColorField.blur();});
    expect(hexColorField.value).toBe('#CCBBAA');
    expect(onChangeSpy).toHaveBeenCalledTimes(2);
  });
```

but I'm getting the warning below
![Screenshot 2020-10-26 173917](https://user-images.githubusercontent.com/5743833/97242873-34a86200-17b2-11eb-9409-622532780d31.png)

I also see it in the stories I have where I have controlled components, after triggering onChange events. I've looked it up in multiple places but haven't got closer to fixing 😆. It's something about calling setState functions while the component is rendering

Reply:
resolved warning above with https://github.com/adobe/react-spectrum/pull/986/commits/813dcf592aba39e34179d39edc260c261c7764c2